opencbm/sys/linux/cbm_module.c

```diff
@@ -700,65 +715,47 @@ static long cbm_unlocked_ioctl(struct file *f,
 			XP_WRITE(0xff);
 			set_data_reverse();
 		}
-		return XP_READ();
+		put_user(XP_READ(), (int *)arg);
+		return 0;
 
 	case CBMCTRL_PP_WRITE:
 		if (data_reverse)
 			set_data_forward();
-		XP_WRITE(arg);
+		XP_WRITE(intarg);
 		return 0;
 
 /* and now the parallel burst-routines */
 
 	case CBMCTRL_PARBURST_READ:
-		return cbm_parallel_burst_read();
+		put_user(cbm_parallel_burst_read(), (int *)arg);
+		return 0;
 
 	case CBMCTRL_PARBURST_WRITE:
-		return cbm_parallel_burst_write(arg);
+		get_user(c, (unsigned char *)arg);
+		return cbm_parallel_burst_write(c);
 
 	case CBMCTRL_PARBURST_READ_TRACK:
-		/* cast arg to structure pointer */
-		user_val = (PARBURST_RW_VALUE *) arg;
-		/* copy the data to the kernel: */
-		if (copy_from_user(&kernel_val,	/* kernel buffer */
```

Comment:
kernel_val is not defined here (and in the lower instruction).

To be honest, I do not understand the logic here completely at the moment. Should kernel_val and val be the same? Can you elaborate a little bit?